Adopt std::span in GIFImageReader

cdumez opened the pull request:
#### 6c6ce2d397a52ce87ebd8e320943148333d5ad8d
<pre>
Adopt std::span in GIFImageReader
<a href="https://bugs.webkit.org/show_bug.cgi?id=272640">https://bugs.webkit.org/show_bug.cgi?id=272640</a>

Reviewed by Darin Adler.

* Source/WebCore/platform/image-decoders/gif/GIFImageDecoder.cpp:
(WebCore::GIFImageDecoder::haveDecodedRow):
* Source/WebCore/platform/image-decoders/gif/GIFImageReader.cpp:
(GIFLZWContext::doLZW):
(GIFFrameContext::decode):
(GIFImageReader::decode):
(GIFImageReader::parse):
* Source/WebCore/platform/image-decoders/gif/GIFImageReader.h:
(GIFImageReader::GIFImageReader):
(GIFImageReader::setData):
(GIFImageReader::globalColormap const):
(GIFImageReader::localColormap const):
(GIFImageReader::data const):
(GIFImageReader::globalColormapSize const): Deleted.
(GIFImageReader::localColormapSize const): Deleted.

Canonical link: <a href="https://commits.webkit.org/277476@main">https://commits.webkit.org/277476@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3182b3dcac0bfd974a273266ffeb7fc58c685a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47733 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50416 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43788 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24400 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38849 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48315 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41208 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20146 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22054 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42399 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5782 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44077 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42821 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52310 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22769 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19097 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46154 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24041 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45188 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24829 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6747 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23761 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/image-decoders/gif/GIFImageReader.h:
```diff
@@ -194,7 +194,7 @@ struct GIFFrameContext {
     }
 
-    bool decode(const unsigned char* data, size_t length, WebCore::GIFImageDecoder* client, bool* frameDecoded);
+    bool decode(std::span<const uint8_t> data, WebCore::GIFImageDecoder* client, bool& frameDecoded);
```
If you are making this out parameter required, you should consider changing the return type to a struct that can represent both outputs.

cdumez replied:
It was already required, so passing a pointer seemed confusing.

darinadler replied:
I've been converting out arguments into return values on functions like this, but it does make the refactoring more risky so I wouldn't insist on it.

darinadler commented on Source/WebCore/platform/image-decoders/gif/GIFImageReader.cpp:
```diff
@@ -334,17 +330,17 @@ bool GIFFrameContext::decode(const unsigned char* data, size_t length, WebCore::
         size_t blockPosition = m_lzwBlocks[m_currentLzwBlock].blockPosition;
         size_t blockSize = m_lzwBlocks[m_currentLzwBlock].blockSize;
-        if (blockPosition + blockSize > length)
+        if (blockPosition + blockSize > data.size())
```
Would be better to write this in a way that won't be confused by overflow.

cdumez committed:
Adopt std::span in GIFImageReader
https://bugs.webkit.org/show_bug.cgi?id=272640

Reviewed by Darin Adler.

* Source/WebCore/platform/image-decoders/gif/GIFImageDecoder.cpp:
(WebCore::GIFImageDecoder::haveDecodedRow):
* Source/WebCore/platform/image-decoders/gif/GIFImageReader.cpp:
(GIFLZWContext::doLZW):
(GIFFrameContext::decode):
(GIFImageReader::decode):
(GIFImageReader::parse):
* Source/WebCore/platform/image-decoders/gif/GIFImageReader.h:
(GIFImageReader::GIFImageReader):
(GIFImageReader::setData):
(GIFImageReader::globalColormap const):
(GIFImageReader::localColormap const):
(GIFImageReader::data const):
(GIFImageReader::globalColormapSize const): Deleted.
(GIFImageReader::localColormapSize const): Deleted.

Canonical link: https://commits.webkit.org/277476@main
